File: backend/app/routers/matching.py
import io

import pandas as pd
from fastapi import APIRouter, Depends, HTTPException, UploadFile
from sqlalchemy import select
from sqlalchemy.orm import Session

from .. import auth
from .. import matching as matching_engine
from ..database import get_db
from ..models import (
    AuditLog,
    Invoice,
    InvoiceStatus,
    MatchEntry,
    MatchResult,
    MatchSource,
    MatchSourceType,
    MatchStatus,
    User,
)
from ..schemas import MatchResultOut, MatchRunSummary, MatchSourceOut

router = APIRouter(prefix="/api/matching", tags=["matching"])

COLUMN_ALIASES = {
    "vendor": {"vendor", "vendor_name", "payee", "supplier", "name"},
    "amount": {"amount", "total", "value"},
    "date": {"date", "entry_date", "transaction_date", "po_date"},
    "reference": {"reference", "po_number", "po_reference", "ref", "check_number", "memo"},
}


def _resolve_columns(df: pd.DataFrame) -> dict[str, str]:
    lower_cols = {c.lower().strip(): c for c in df.columns}
    resolved = {}
    for canonical, aliases in COLUMN_ALIASES.items():
        for alias in aliases:
            if alias in lower_cols:
                resolved[canonical] = lower_cols[alias]
                break
    if "vendor" not in resolved or "amount" not in resolved:
        raise HTTPException(
            status_code=422,
            detail=f"CSV must include a vendor/payee column and an amount column. Found columns: {list(df.columns)}",
        )
    return resolved


@router.post("/sources", response_model=MatchSourceOut, status_code=201)
def upload_match_source(
    source_type: str,
    file: UploadFile,
    db: Session = Depends(get_db),
    current_user: User = Depends(auth.get_current_user),
):
    if source_type not in (MatchSourceType.PO.value, MatchSourceType.BANK.value):
        raise HTTPException(status_code=422, detail="source_type must be 'po' or 'bank'")

    raw = file.file.read()
    try:
        df = pd.read_csv(io.BytesIO(raw))
    except Exception as exc:
        raise HTTPException(status_code=422, detail=f"Could not parse CSV: {exc}") from exc

    cols = _resolve_columns(df)

    source = MatchSource(
        org_id=current_user.org_id,
        name=file.filename or "upload.csv",
        source_type=MatchSourceType(source_type),
    )
    db.add(source)
    db.flush()

    for _, row in df.iterrows():
        entry_date = None
        if "date" in cols and pd.notna(row.get(cols["date"])):
            parsed = pd.to_datetime(row[cols["date"]], errors="coerce")
            entry_date = parsed.date() if pd.notna(parsed) else None

        amount = None
        if pd.notna(row.get(cols["amount"])):
            try:
                amount = float(str(row[cols["amount"]]).replace("$", "").replace(",", ""))
            except ValueError:
                amount = None

        db.add(
            MatchEntry(
                source_id=source.id,
                vendor=str(row.get(cols["vendor"], "") or ""),
                amount=amount,
                entry_date=entry_date,
                reference=str(row.get(cols.get("reference"), "") or "") if "reference" in cols else "",
                raw_row=row.astype(str).to_dict(),
            )
        )

    db.add(
        AuditLog(
            org_id=current_user.org_id,
            user_id=current_user.id,
            action="match_source_uploaded",
            actor=current_user.email,
            details={"source_type": source_type, "rows": len(df)},
        )
    )
    db.commit()
    db.refresh(source)

    result = MatchSourceOut.model_validate(source)
    result.entry_count = len(df)
    return result


@router.get("/sources", response_model=list[MatchSourceOut])
def list_match_sources(
    db: Session = Depends(get_db),
    current_user: User = Depends(auth.get_current_user),
):
    stmt = select(MatchSource).where(MatchSource.org_id == current_user.org_id)
    sources = db.scalars(stmt).all()
    out = []
    for s in sources:
        item = MatchSourceOut.model_validate(s)
        item.entry_count = len(s.entries)
        out.append(item)
    return out


@router.post("/run", response_model=MatchRunSummary)
def run_matching(
    db: Session = Depends(get_db),
    current_user: User = Depends(auth.get_current_user),
):
    entries = db.scalars(
        select(MatchEntry).join(MatchSource).where(MatchSource.org_id == current_user.org_id)
    ).all()
    candidates = [
        matching_engine.MatchCandidateEntry(
            id=e.id, vendor=e.vendor, amount=e.amount, entry_date=e.entry_date, reference=e.reference
        )
        for e in entries
    ]

    invoices = db.scalars(
        select(Invoice).where(
            Invoice.org_id == current_user.org_id,
            Invoice.status.in_(
                [InvoiceStatus.EXTRACTED, InvoiceStatus.NEEDS_REVIEW, InvoiceStatus.APPROVED]
            ),
        )
    ).all()

    counts = {"matched": 0, "partial": 0, "unmatched": 0}
    for invoice in invoices:
        outcome = matching_engine.find_best_match(
            invoice.vendor_name, invoice.total, invoice.invoice_date, invoice.po_reference, candidates
        )
        db.add(
            MatchResult(
                invoice_id=invoice.id,
                match_entry_id=outcome.entry_id,
                status=MatchStatus(outcome.status),
                score=outcome.score,
                reasoning=outcome.reasoning,
            )
        )
        db.add(
            AuditLog(
                org_id=current_user.org_id,
                user_id=current_user.id,
                invoice_id=invoice.id,
                action="match_evaluated",
                actor=current_user.email,
                details={"status": outcome.status, "score": outcome.score, "reasoning": outcome.reasoning},
            )
        )
        counts[outcome.status] += 1

    db.commit()

    return MatchRunSummary(
        invoices_evaluated=len(invoices),
        matched=counts["matched"],
        partial=counts["partial"],
        unmatched=counts["unmatched"],
    )


@router.get("/results", response_model=list[MatchResultOut])
def list_match_results(
    db: Session = Depends(get_db),
    current_user: User = Depends(auth.get_current_user),
):
    stmt = (
        select(MatchResult)
        .join(Invoice, MatchResult.invoice_id == Invoice.id)
        .where(Invoice.org_id == current_user.org_id)
        .order_by(MatchResult.created_at.desc())
    )
    return db.scalars(stmt).all()

File: backend/src/routes/matching.js
import { Router } from "express";
import { parse } from "csv-parse/sync";
import { requireAuth } from "../auth.js";
import { requireActivePlan } from "../plan.js";
import * as matchingEngine from "../matching.js";
import { upload } from "../storage.js";
import { AuditLog, Invoice, MatchEntry, MatchResult, MatchSource } from "../models/index.js";
import { serializeMatchResult, serializeMatchSource } from "../serializers.js";

const router = Router();

const COLUMN_ALIASES = {
  vendor: ["vendor", "vendor_name", "payee", "supplier", "name"],
  amount: ["amount", "total", "value"],
  date: ["date", "entry_date", "transaction_date", "po_date"],
  reference: ["reference", "po_number", "po_reference", "ref", "check_number", "memo"],
};

function resolveColumns(headers) {
  const lowerToActual = new Map(headers.map((h) => [h.toLowerCase().trim(), h]));
  const resolved = {};
  for (const [canonical, aliases] of Object.entries(COLUMN_ALIASES)) {
    for (const alias of aliases) {
      if (lowerToActual.has(alias)) {
        resolved[canonical] = lowerToActual.get(alias);
        break;
      }
    }
  }
  if (!resolved.vendor || !resolved.amount) {
    return null;
  }
  return resolved;
}

function parseAmount(raw) {
  if (raw === undefined || raw === null || raw === "") return null;
  const cleaned = String(raw).replace(/[$,]/g, "").trim();
  const n = Number(cleaned);
  return Number.isFinite(n) ? n : null;
}

function parseDateToISO(raw) {
  if (!raw) return null;
  const value = String(raw).trim();
  const iso = value.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (iso) return `${iso[1]}-${iso[2]}-${iso[3]}`;
  const slash = value.match(/^(\d{1,2})\/(\d{1,2})\/(\d{2,4})$/);
  if (slash) {
    let [, m, d, y] = slash;
    if (y.length === 2) y = `20${y}`;
    if (Number(m) > 12 && Number(d) <= 12) [m, d] = [d, m];
    return `${y}-${m.padStart(2, "0")}-${d.padStart(2, "0")}`;
  }
  return null;
}

router.post("/api/matching/sources", requireAuth, requireActivePlan, upload.single("file"), async (req, res, next) => {
  try {
    const sourceType = req.query.source_type;
    if (!["po", "bank", "receiving"].includes(sourceType)) {
      return res.status(422).json({ detail: "source_type must be 'po', 'bank', or 'receiving'" });
    }
    if (!req.file) {
      return res.status(422).json({ detail: "A file upload is required." });
    }

    const fs = await import("node:fs/promises");
    const raw = await fs.readFile(req.file.path);
    await fs.rm(req.file.path, { force: true });

    let records;
    try {
      records = parse(raw, { columns: true, skip_empty_lines: true, trim: true });
    } catch (exc) {
      return res.status(422).json({ detail: `Could not parse CSV: ${exc.message}` });
    }

    const headers = records.length ? Object.keys(records[0]) : [];
    const cols = resolveColumns(headers);
    if (!cols) {
      return res.status(422).json({
        detail: `CSV must include a vendor/payee column and an amount column. Found columns: ${JSON.stringify(headers)}`,
      });
    }

    const source = await MatchSource.create({ orgId: req.currentUser.orgId, name: req.file.originalname || "upload.csv", sourceType });

    await MatchEntry.bulkCreate(
      records.map((row) => ({
        sourceId: source.id,
        vendor: row[cols.vendor] || "",
        amount: parseAmount(row[cols.amount]),
        entryDate: cols.date ? parseDateToISO(row[cols.date]) : null,
        reference: cols.reference ? row[cols.reference] || "" : "",
        rawRow: row,
      }))
    );

    await AuditLog.create({
      orgId: req.currentUser.orgId,
      userId: req.currentUser.id,
      action: "match_source_uploaded",
      actor: req.currentUser.email,
      details: { source_type: sourceType, rows: records.length },
    });

    res.status(201).json(serializeMatchSource(source, records.length));
  } catch (err) {
    next(err);
  }
});

router.get("/api/matching/sources", requireAuth, requireActivePlan, async (req, res, next) => {
  try {
    const sources = await MatchSource.findAll({ where: { orgId: req.currentUser.orgId } });
    const out = await Promise.all(
      sources.map(async (s) => serializeMatchSource(s, await MatchEntry.count({ where: { sourceId: s.id } })))
    );
    res.json(out);
  } catch (err) {
    next(err);
  }
});

router.delete("/api/matching/sources/:id", requireAuth, requireActivePlan, async (req, res, next) => {
  try {
    const source = await MatchSource.findOne({ where: { id: req.params.id, orgId: req.currentUser.orgId } });
    if (!source) return res.status(404).json({ detail: "Source not found" });

    const entryCount = await MatchEntry.count({ where: { sourceId: source.id } });

    await AuditLog.create({
      orgId: req.currentUser.orgId,
      userId: req.currentUser.id,
      action: "match_source_deleted",
      actor: req.currentUser.email,
      details: { name: source.name, source_type: source.sourceType, entry_count: entryCount },
    });

    // Cascades to this source's MatchEntry rows (see models/index.js). Any
    // MatchResult that was once matched against one of them keeps its own
    // history -- match_entry_id just goes null -- since a match result is a
    // record of a past evaluation, not something this source still owns.
    await source.destroy();
    res.json({ ok: true });
  } catch (err) {
    next(err);
  }
});

// Runs three-way matching (PO + goods receipt + invoice) as soon as the org
// has any goods receipts on file, and the original two-way match otherwise.
// Inferred from the uploaded sources rather than asked for via a mode flag:
// uploading receipts is already an unambiguous statement that you want them
// checked, and an org with none has nothing to run the third leg against.
// The mode actually used comes back in the response either way, so it's
// visible rather than a silent behavior change.
router.post("/api/matching/run", requireAuth, requireActivePlan, async (req, res, next) => {
  try {
    const sources = await MatchSource.findAll({
      where: { orgId: req.currentUser.orgId },
      attributes: ["id", "sourceType"],
    });
    const entries = await MatchEntry.findAll({ where: { sourceId: sources.map((s) => s.id) } });

    const typeBySourceId = new Map(sources.map((s) => [s.id, s.sourceType]));
    const toCandidate = (e) => ({
      id: e.id,
      vendor: e.vendor,
      amount: e.amount,
      entryDate: e.entryDate,
      reference: e.reference,
    });
    const candidates = entries.map(toCandidate);
    const poCandidates = entries.filter((e) => typeBySourceId.get(e.sourceId) === "po").map(toCandidate);
    const receivingCandidates = entries.filter((e) => typeBySourceId.get(e.sourceId) === "receiving").map(toCandidate);

    const threeWay = receivingCandidates.length > 0;

    const invoices = await Invoice.findAll({
      where: { orgId: req.currentUser.orgId, status: ["extracted", "needs_review", "approved"] },
    });

    const counts = { matched: 0, partial: 0, unmatched: 0 };
    const outcomeCounts = { matched: 0, no_receipt: 0, no_po: 0, unmatched: 0 };

    for (const invoice of invoices) {
      const outcome = threeWay
        ? matchingEngine.findThreeWayMatch(
            invoice.vendorName,
            invoice.total,
            invoice.invoiceDate,
            invoice.poReference,
            poCandidates,
            receivingCandidates
          )
        : matchingEngine.findBestMatch(
            invoice.vendorName,
            invoice.total,
            invoice.invoiceDate,
            invoice.poReference,
            candidates
          );

      await MatchResult.create({
        invoiceId: invoice.id,
        matchEntryId: outcome.entryId,
        receivingEntryId: outcome.receivingEntryId ?? null,
        status: outcome.status,
        threeWayOutcome: outcome.threeWayOutcome ?? null,
        score: outcome.score,
        reasoning: outcome.reasoning,
      });
      await AuditLog.create({
        orgId: req.currentUser.orgId,
        userId: req.currentUser.id,
        invoiceId: invoice.id,
        action: "match_evaluated",
        actor: req.currentUser.email,
        details: {
          mode: threeWay ? "three_way" : "two_way",
          status: outcome.status,
          three_way_outcome: outcome.threeWayOutcome ?? null,
          score: outcome.score,
          reasoning: outcome.reasoning,
        },
      });
      counts[outcome.status] += 1;
      if (threeWay) outcomeCounts[outcome.threeWayOutcome] += 1;
    }

    res.json({
      mode: threeWay ? "three_way" : "two_way",
      invoices_evaluated: invoices.length,
      matched: counts.matched,
      partial: counts.partial,
      unmatched: counts.unmatched,
      // Only meaningful on a three-way run -- null (rather than a set of
      // zeroes) on a two-way one, so the UI can tell "not evaluated" apart
      // from "evaluated, found none".
      three_way: threeWay ? outcomeCounts : null,
    });
  } catch (err) {
    next(err);
  }
});

router.get("/api/matching/results", requireAuth, requireActivePlan, async (req, res, next) => {
  try {
    const results = await MatchResult.findAll({
      include: [{ model: Invoice, attributes: [], where: { orgId: req.currentUser.orgId }, required: true }],
      order: [["createdAt", "DESC"]],
    });
    res.json(results.map(serializeMatchResult));
  } catch (err) {
    next(err);
  }
});

export default router;
